### PR TITLE
Adding regular 2d cartesian interpolation for the injectivity simulation. 

### PR DESCRIPTION
**Pull request by GitPaean**
Basically, it is for the keywords `PLYMWINJ`, `SKPRWAT` and `SKPRPOLY` introduced in https://github.com/OPM/opm-common/pull/520

Some small adjustments were introduced to make some code work with the `DynamicEvaluation` from https://github.com/OPM/opm-material/pull/302

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/material/densead/Evaluation.hpp [hunk 1]
```diff
@@ -538,9 +538,7 @@ Evaluation<ValueType, numVars> operator+(const RhsValueType& a, const Evaluation
 template <class RhsValueType, class ValueType, int numVars>
 Evaluation<ValueType, numVars> operator-(const RhsValueType& a, const Evaluation<ValueType, numVars>& b)
 {
-    Evaluation<ValueType, numVars> result(a);
-    result -= b;
-    return result;
+    return -(b - a);
```
Comment (atgeirr) [inline opm/material/densead/Evaluation.hpp]: This change looks odd to me: what is the benefit of this change?

Comment (GitPaean) [reply opm/material/densead/Evaluation.hpp]: I think ` Evaluation<ValueType, numVars> result(a);` does not work for `DynamicEvaluation`. 

With this change we reuse ` Evaluation operator-(const RhsValueType& other) const`.

Comment (GitPaean) [reply opm/material/densead/Evaluation.hpp]: But it is slower. Maybe we should look into other options.

File: tests/test_2dtables.cpp [hunk 2]
```diff
@@ -132,6 +133,37 @@ createUniformXTabulatedFunction2(Fn& f)
...
 }
 
+
+template <class Fn>
+std::shared_ptr<Opm::XYTabulated2DFunction<Scalar> >
+createXYTabulated2DFunction(Fn& f)
+{
+    const Scalar xMin = -2.0;
+    const Scalar xMax = 3.0;
+    const Scalar m = 50;
```
Comment (atgeirr) [inline tests/test_2dtables.cpp]: Should be integer, not Scalar.

File: opm/material/common/XYTabulated2DFunction.hpp [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+                          const bool y_extrapolate = false)
+       : xPos_(x_pos)
+       , yPos_(y_pos)
+       , samples_(data)
+       , xExtrapolate_(x_extrapolate)
+       , yExtrapolate_(y_extrapolate)
+    {
+        // make sure the size is correct
+        if (numX() != samples_.size()) {
+            throw std::runtime_error("numX() is not equal with the number of rows of the sample points");
```
Comment (atgeirr) [inline opm/material/common/XYTabulated2DFunction.hpp]: "... not equal TO ..."

File: opm/material/common/XYTabulated2DFunction.hpp [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+        }
+
+        for (unsigned ix = 0; ix < numX(); ++ix) {
+            if (samples_[ix].size() != numY()) {
+                std::ostringstream oss;
+                oss << "the " << ix << "th row of the smaples_ has different number of data from numY() ";
```
Comment (atgeirr) [inline opm/material/common/XYTabulated2DFunction.hpp]: smaples_ -> sample points

File: opm/material/common/XYTabulated2DFunction.hpp [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            return 0;
+        else if (x >= xMax())
+                return numX() - 2;
+        else {
+            assert(numX() >= 3);
+
+            // bisection
+            unsigned lowerIdx = 0;
+            unsigned upperIdx = numX() - 1;
+            while (lowerIdx + 1 < upperIdx) {
```
Comment (atgeirr) [inline opm/material/common/XYTabulated2DFunction.hpp]: Why not use std::lower_bound()?

Also: Logic is identical between x and y, use a single function, implement using common function (can keep the existing functions, and let them call a new common method).

**Comment by dr-robertk** [comment]
Are you guys reinventing the wheel here? Why not just use YaspGrid from the dune-grid module? It's already there and it works.

**Comment by atgeirr** [comment]
> Are you guys reinventing the wheel here? Why not just use YaspGrid from the dune-grid module? It's already there and it works.

The PR title is misleading, it is a 2d interpolation code, not a grid in the Dune sense.

File: opm/material/common/XYTabulated2DFunction.hpp (hunk 5 above)
Comment (GitPaean) [reply opm/material/common/XYTabulated2DFunction.hpp]: I changed to use `std::lower_bound()`. I have the impression that sometimes the meaning of some functions from algorithm is not `that` clear.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by dr-robertk** [comment]
@atgeirr: I see.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by andlaus** [comment]
besides some minor style issues (which I'm gonna fix myself, no worries), please do *not* manually change auto-generated code (`DynamicEvaluation.hpp`, `Evaluation.hpp`, etc.). Instead, modify the code generator `bin/genEvalSpecializations.py`, run it and put the the output into a separate commit...

**Comment by andlaus** [comment]
maybe jenkins could be amended to check that no auto-generated code is manually modified: run the code generator at the build and check that nothing changed. this could be done via a "jenkins pre-check" script inside the repository, so that the functionality could be quite generic and would not clutter the actual jenkins scripts. what do you think @akva2? (I'll also implement this for you if you want.)

**Comment by andlaus** [comment]
@GitPaean, @atgeirr: will you change the code generator? if so, what's the ETA?

**Comment by GitPaean** [comment]
> @GitPaean, @atgeirr: will you change the code generator? if so, what's the ETA?

I did not receive your message clearly. I thought you offered to do it since it is something you are the most familiar with. 

I will take a look and let you know if I have any problem.

**Comment by andlaus** [comment]
> I did not receive your message clearly. I thought you offered to do it since it is something you are the most familiar with.

I meant to say that I'll do these coding style issues. I could also do the code-gen, but I think it is a good didactic experience for you.

**Pull request merged**